lib/auth/init_test.go

```diff
@@ -154,158 +154,211 @@ func TestBadIdentity(t *testing.T) {
 	require.IsType(t, trace.BadParameter(""), err)
 }
 
-// TestAuthPreference ensures that the act of creating an AuthServer sets
-// the AuthPreference (type and second factor) on the backend.
-func TestAuthPreference(t *testing.T) {
-	conf := setupConfig(t)
-	conf.AuthPreference = newU2FAuthPreferenceFromConfigFile(t)
-	as, err := Init(conf)
-	require.NoError(t, err)
-	defer as.Close()
-
-	cap, err := as.GetAuthPreference()
-	require.NoError(t, err)
-	require.Empty(t, resourceDiff(cap, conf.AuthPreference))
+type testDynamicallyConfigurableParams struct {
+	withDefaults, withConfigFile, withAnotherConfigFile func(*InitConfig) types.ResourceWithOrigin
+	setDynamic                                          func(*Server)
+	getStored                                           func(*Server) types.ResourceWithOrigin
 }
 
-func TestAuthPreferenceInitFromConfigFileToDefault(t *testing.T) {
-	// Simulate a server with auth preference from config file.
-	var err error
-	conf := setupConfig(t)
-	conf.AuthPreference, err = types.NewAuthPreferenceFromConfigFile(types.AuthPreferenceSpecV2{
-		SecondFactor: constants.SecondFactorOff,
+func testDynamicallyConfigurable(t *testing.T, p testDynamicallyConfigurableParams) {
+	initAuthServer := func(conf InitConfig) *Server {
```

Comment:
Should `initAuthServer` take in the `t *testing.T` parameter?

Reply:
Well, I don't know, should it? :) The only relevant `t` is captured by the closure so the code apparently works as is.

Reply:
Sorry that was a vague comment as I wasn't positive about the following. `t.Run` spawns a child `testing.T`, so the `t.Cleanup` in `initAuthServer` doesn't run until `testDynamicallyConfigurable` returns. If you pass the `testing.T` spawned by `t.Run`, the cleanup will occur immediately after `t.Run` concludes. It's a small performance improvement that will matter more in other tests.

Reply:
please pass `testing.T` as an argument here
the closure works, but any failure will be associated with the top-level test, not the subtest
passing a subtest-specific `t` will make the failure clearer